[Validator] Expression language allow null

mpiot opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | yes
| Tickets       | 
| License       | MIT
| Doc PR        | 

Like other validators, ExpressionLanguageSyntax validator should be ignored when the value is null or receive an empty string, add NotNull or NotBlank if needed.

mpiot committed:
[Constrainst] Expression language allow null

Andrej-in-ua commented:
Looks like correct fix. Thanks!

mpiot commented:
Thanks for your review 🙂

derrabus commented:
I agree that the validator should behave as you're suggesting. However, changing its behavior is technically a BC break, so we should probably add a config flag and a deprecation layer for your change.

mpiot committed:
Add allowNullAndEmptyString option

mpiot commented:
@derrabus I've added an `allowNullAndEmptyString` option, and add a deprecation on its usage for 5.4.
Added some tests to: with and without the option.

For the deprecation I don't really know ho to add it correctly. For the moment just add a Deprecation when using `allowNullAndEmptyString` but for the 5.4 version that do not exists, it trigger a Deprecation that we can't solve.

derrabus commented on src/Symfony/Component/Validator/Constraints/ExpressionLanguageSyntax.php:
```diff
@@ -31,14 +31,20 @@ class ExpressionLanguageSyntax extends Constraint
         $this->allowedVariables = $allowedVariables ?? $this->allowedVariables;
+        $this->allowNullAndEmptyString = $allowNullAndEmptyString ?? $this->allowNullAndEmptyString;
+
+        if (isset($options['allowNullAndEmptyString'])) {
+            trigger_deprecation('symfony/validator', '5.4', sprintf('The "allowNullAndEmptyString" option of the "%s" constraint is deprecated.', self::class));
+        }
```
This parameter is our forward compatibility layer and it allows us to opt-in to the behavior we want to have in 6.0. An application that does not explicitly enable that new behavior will potentially break when upgrading to Symfony 6.

What we should deprecate is not explicitly setting it to `true`.

```suggestion
        if (!$this->allowNullAndEmptyString) {
            trigger_deprecation('symfony/validator', '5.4', 'Not setting "allowNullAndEmptyString" of the "%s" constraint to "true" is deprecated.', __CLASS__);
        }
```

derrabus reviewed (changes requested):
Sorry for stalling here. I have a comment about the deprecation layer. We should make sure that the deprecation supports our upgrade path to 6.0.

Also, when testing the deprecated behavior, please make sure to add it to the legacy group:

```php
/**
 * @group legacy
 */
public function testSomeOldBehavior()
{
```

This way, the deprecation will not cause the test to fail and we know which tests can be removed on 6.0.

mpiot committed:
Add test groups and change deprecation message + trigger

HeahDude commented on src/Symfony/Component/Validator/Constraints/ExpressionLanguageSyntax.php:
```diff
@@ -31,14 +31,20 @@ class ExpressionLanguageSyntax extends Constraint
+        $this->allowNullAndEmptyString = $allowNullAndEmptyString ?? $this->allowNullAndEmptyString;
+
+        if (!$this->allowNullAndEmptyString) {
+            trigger_deprecation('symfony/validator', '5.4', 'Not setting "allowNullAndEmptyString" of the "%s" constraint to "true" is deprecated.', __CLASS__);
```
I propose to be more explicit on the upgrade path here, something like `Validating empty expressions with "%s" constraint is deprecated. Set "allowNullAndEmptyString" option to "true" instead and add explicit constraint like NotNull or NotBlank.'`

mpiot replied:
@HeahDude I think you're right, I"ve just base the message on other validator that do the same. But be more explicit is maybe better for final user.

HeahDude commented on src/Symfony/Component/Validator/Constraints/ExpressionLanguageSyntaxValidator.php:
```diff
@@ -39,6 +39,10 @@ public function validate($expression, Constraint $constraint): void
         }
 
+        if (true === $constraint->allowNullAndEmptyString && (null === $expression || '' === $expression)) {
+            return;
```
The deprecation should be triggered here instead of the constraint constructor to ensure runtime context, otherwise it may be lost in the app cache warm up, something like:
```php
if (null === $expression || '' === $expression) {
    if ($constraint->allowNullAndEmptyString) {
        return;
    }

    trigger_deprecation(...);
}

mpiot replied:
I'll wait for other comments about it, because all triggered deprecations in Constraints are in the Constraint class and not in the Validator (eg: Length, Range)

HeahDude replied:
I would say that it should never be the case, otherwise it may end up as symfony/symfony-docs#14785 (ref https://github.com/symfony/symfony-docs/pull/14785#discussion_r563106869).

HeahDude reviewed (commented):
Thanks for that PR. What about fixing https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/Validator/Constraints/ExpressionValidator.php#L42 as well, could be done in the same PR I guess.

mpiot committed:
More explicit deprecation message

jderusse commented:
> However, changing its behavior is technically a BC break, so we should probably add a config flag and a deprecation layer for your change.

I disagree with this.

This validator is the **only one** that throws an exception when the value is null:

- https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/Validator/Constraints/BicValidator.php#L67
- https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/Validator/Constraints/UrlValidator.php#L51
- https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/Validator/Constraints/DateValidator.php#L45
- https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/Validator/Constraints/FileValidator.php#L50
- https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/Validator/Constraints/IbanValidator.php#L149
- https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/Validator/Constraints/IsbnValidator.php#L39
- https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/Validator/Constraints/IsinValidator.php#L35
- https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/Validator/Constraints/IssnValidator.php#L38
- https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/Validator/Constraints/JsonValidator.php#L32
- https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/Validator/Constraints/LuhnValidator.php#L46
- https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/Validator/Constraints/TimeValidator.php#L45
- https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/Validator/Constraints/TypeValidator.php#L64
- https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/Validator/Constraints/UlidValidator.php#L36
- https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/Validator/Constraints/UuidValidator.php#L71
- https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/Validator/Constraints/CountValidator.php#L33
- https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/Validator/Constraints/EmailValidator.php#L55
- https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/Validator/Constraints/ImageValidator.php#L43
- https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/Validator/Constraints/RangeValidator.php#L43
- https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/Validator/Constraints/RegexValidator.php#L36
- https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/Validator/Constraints/ChoiceValidator.php#L42
- https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/Validator/Constraints/IsTrueValidator.php#L32
- https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/Validator/Constraints/LengthValidator.php#L33
- https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/Validator/Constraints/LocaleValidator.php#L36
- https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/Validator/Constraints/UniqueValidator.php#L33
- https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/Validator/Constraints/CountryValidator.php#L36

IMHO, because of such inconsistency with the rest of the code, this is a bug and could not be an expected behavior.

derrabus commented:
> IMHO, because of such inconsistency with the rest of the code, this is a bug and could not be an expected behavior.

I agree with that assessment. However, it's the current behavior of stable Symfony releases, even if it's an unexpected one. If we simply change it, we might break existing validation logic of apps.

We've had a similar situation with the `LengthValidator` where it did not no-op on empty strings like all other validators do. This has been solved in a simlar way.

fabpot commented:
Would be great to have this in 5.4.

mpiot commented:
I've rebase on 5.4 branch, but change nothing in the code (divergence in the discussion).

jderusse commented on src/Symfony/Component/Validator/Constraints/ExpressionLanguageSyntax.php:
```diff
@@ -31,14 +31,20 @@ class ExpressionLanguageSyntax extends Constraint
+    public $allowNullAndEmptyString = false;
 
-    public function __construct(array $options = null, string $message = null, string $service = null, array $allowedVariables = null, array $groups = null, $payload = null)
+    public function __construct(array $options = null, string $message = null, string $service = null, array $allowedVariables = null, bool $allowNullAndEmptyString = null, array $groups = null, $payload = null)
```
Shouldn't we move the parameter at the end for BC? (I wonder if it matters for annotations :thinking:)

jderusse reviewed (approved):
Ok for me (just a minor comment about BC), 
Then in 6.1 we could deprecate that new param

fabpot commented:
6.0 cannot contain deprecations, any deprecation should be in 6.1 instead.

mpiot commented:
Can we can add the deprecation message in 5.4, and remove it and the alternative logic for 6.0 branch (2nd PR) ?
As version 6 is the same as version 5.4 without deprecations.

stof commented:
it is too late to add new deprecations in 5.4. We are already in RC. the feature freeze was more than 1 month ago.

nicolas-grekas commented:
I think adding argument `$allowNullAndEmptyString` does not make sense.

Instead of the current approach which is trying to fix a broken behavior of `ExpressionLanguageSyntax`, I propose to deprecate it and replace it with a new constraint named `ExpressionSyntax`.

@mpiot up to implement this?

mpiot commented:
@nicolas-grekas yes, I can create a new validator instead of update this one and deprecate the current validator.

Just a question about the part:
public const EXPRESSION_LANGUAGE_SYNTAX_ERROR = '1766a3f3-ff03-40eb-b053-ab7aa23d988a';

What is the way to “create” this part  ? UUID ?

nicolas-grekas commented:
Replaced by #45623
